doc/high-frequency-telemetry/high-frequency-telemetry-hld.md

```diff
@@ -0,0 +1,706 @@
+
+- Receivers: Accepts OTLP metrics via OTLP/gRPC protocol
+- Processors: Batches metrics for efficient transmission
+- Exporters: Forwards metrics to backend databases for storage and visualization
+
+``` mermaid
+
+flowchart TD;
+    CS[CounterSyncd] -->|OTLP/gRPC| R
+    subgraph OpenTelemetry Collector
+        R[Receivers] --> P[Processors] --> E[Exporters]
+    end
+    BS@{ shape: fork, label: "Backend service" }
+    subgraph Backend Service
+        IDB[InfluxDB]
+        PM[Prometheus]
+        OTH[Other Options]
+    end
+    E --> |HTTP/API|BS
+    BS --> IDB
+    BS --> PM
+    BS --> OTH
+
+```
+
+For further details on OpenTelemetry and OpenTelemetry Collector, please refer to the official documentation:
+
+- [What is OpenTelemetry?](https://opentelemetry.io/docs/what-is-opentelemetry/)
+- [OpenTelemetry Collector](https://opentelemetry.io/docs/collector/)
+
+### 7.2. Data format
+
+We will use IPFIX as the report format, with all numbers in the IPFIX message in network-order (Big-endian).
+
+For more information on IPFIX, refer to the following resources:
+
+- [Specification of the IP Flow Information Export (IPFIX) Protocol for the Exchange of Flow Information](https://datatracker.ietf.org/doc/html/rfc7011)
+- [IP Flow Information Export (IPFIX) Entities](https://www.iana.org/assignments/ipfix/ipfix.xhtml)
+
+#### 7.2.1. IPFIX header
+
+The `Version` and `Observation Domain ID` fields of the IPFIX header are identical for each IPFIX message.
+
+``` mermaid
+
+---
+title: stream message IPFIX header
+---
+packet-beta
+0-15: "Version = 0x000a"
+16-31: "Message Length = (16 + payload) bytes"
+32-63: "Export Timestamp: Second"
+64-95: "Sequence Number = 0, start from 0 and incremental sequence counter modulo 2^32"
+96-127: "Observation Domain ID = 0, always 0"
+
+```
+
+#### 7.2.2. IPFIX template
+
+``` mermaid
+
+---
+title: stream message of IPFIX template
+---
+packet-beta
+0-15: "Set ID = 2"
+16-31: "Set Length = (12 + Number of Stats * 8) bytes"
+32-47: "Template ID = > 256 configured"
+48-63: "Number of Fields = 1 + Number of stats"
+64-79: "Element ID=observationTimeNanoseconds (325)"
+80-95: "Field length = 8 bytes"
+96-96: "1"
+97-111: "Element ID = Object index for the stats 1"
+112-127: "Field Length = 8 bytes"
+128-159: "Enterprise Number = SAI TYPE ID + SAI STATS ID for the stats 1"
+160-191: "..."
+192-192: "1"
+193-207: "Element ID = Object index for the stats N"
+208-223: "Field Length = 8 bytes"
+224-255: "Enterprise Number = SAI TYPE ID + SAI STATS ID for the stats N"
+
+```
+
+- For high-frequency counters, the native IPFIX timestamp unit of seconds is insufficient. Therefore, we introduce an additional element, `observationTimeNanoseconds`, for each record to meet our requirements.
+- The element ID of IPFIX is derived from the object index. For example, for `Ethernet5`, the element ID will be `0x5 | 0x8000 = 0x8005`, where `0x8000` indicates that the enterprise bit is set to 1.
```

Comment:
Please describe this in detail. Some objects like buffer_pool, buffer_profile etc. do not have a user visible index.
Please also describe how this element-id to object-id translation is to be done in _counter syncd_ and if there are any information shared between __orchagent__ and _counter syncd_ for this.

Reply:
Thanks for your constructive suggestion and reminder. I didn't notice there is object, such as buffer_pool, without a user visible index.
So, I re-design this part. Please check the commit: https://github.com/sonic-net/SONiC/pull/1795/commits/f3aef298b71b176879f5cf417ad8283d282189d3

In our new design, the counter syncd can map from element ID to object name via session tables in state db. And if we want to get the mapping between object name to SAI OID, there is already a name map in counter db. We can directly leverage it.
![image](https://github.com/user-attachments/assets/b39c9939-142f-47db-b889-1113b1bbdf82)
![image](https://github.com/user-attachments/assets/2b9d927d-0515-4e1b-88e4-8b292e4c0e1b)